### PR TITLE
Upgrade Kubernetes to v1.30.7

### LINKED ISSUE #2457: Support Kubernetes v1.28 - v1.31
### What you would like to be added?

Similar to Training Operator, we should support Kubernetes v1.28 - v1.31 for Katib: https://github.com/kubeflow/training-operator/issues/2291

By following these PRs, we can see what changes are required for Katib:
- https://github.com/kubeflow/training-operator/pull/2332
- https://github.com/kubeflow/training-operator/pull/2330

/good-first-issue

### Love this feature?

Give it a 👍 We prioritize the features with most 👍

### PR DESCRIPTION
**Pull request by Electronic-Waste**
**What this PR does / why we need it**:

- Kubernetes upgrade to v1.30.7
- controller-runtime upgrade to v0.18.5
- Adapt controllers to use controller-runtime generics API
- Adapt code generation to new kube_codegen.sh script

**Which issue(s) this PR fixes** _(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)_:
Fixes #2457

**Checklist:**

- [ ] [Docs](https://www.kubeflow.org/docs/components/katib/) included if any changes are user facing

### CODE REVIEW

**Comment by Electronic-Waste** [comment]
PTAL if you have time👀 @kubeflow/wg-automl-leads

**Review by andreyvelich — COMMENTED** [review]
Thank you for this @Electronic-Waste!
I left my initial comments.
/assign @kubeflow/wg-training-leads @helenxie-bit

File: .github/workflows/e2e-test-darts-cifar10.yaml [hunk 1]
```diff
@@ -33,6 +33,6 @@ jobs:
     strategy:
       fail-fast: false
       matrix:
-        kubernetes-version: ["v1.27.11", "v1.28.7", "v1.29.2"]
+        kubernetes-version: ["v1.28.7", "v1.29.2", "v1.30.7"]
```
[the remark this answers is not in the document]
Comment (Electronic-Waste) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: Yeah, I'd like to. But I'm busy this week. If it's not urgent, I'll raise another PR to migrate to k8s 1.31 next week.

File: go.mod [hunk 2]
```diff
@@ -1,6 +1,8 @@
 module github.com/kubeflow/katib
 
-go 1.22
+go 1.22.0
+
+toolchain go1.23.2
```
[the remark this answers is not in the document]
Comment (Electronic-Waste) [reply go.mod:5]: I guess, it is added automatically by my go environment. I'll remove it.

File: hack/update-openapigen.sh [hunk 3]
```diff
@@ -20,41 +20,26 @@ set -o pipefail
...
-  # shellcheck disable=SC2140
-  go run "${CODEGEN_PKG}/cmd/openapi-gen/main.go" \
-    --go-header-file "${PROJECT_ROOT}/hack/boilerplate/boilerplate.go.txt" \
-    --input-dirs "github.com/kubeflow/katib/pkg/apis/controller/common/$VERSION","github.com/kubeflow/katib/pkg/apis/controller/experiments/$VERSION","github.com/kubeflow/katib/pkg/apis/controller/suggestions/${VERSION}","github.com/kubeflow/katib/pkg/apis/controller/trials/${VERSION}" \
-    --output-package "github.com/kubeflow/katib/pkg/apis/${VERSION}" \
-    "$@"
-
-  echo "Generating swagger file for ${VERSION} ..."
-  go run "${PROJECT_ROOT}/hack/swagger/main.go" "${VERSION}-${SWAGGER_VERSION}" "${VERSION}" >"${SWAGGER_CODEGEN_FILE}"
+  kube::codegen::gen_openapi \
```
[the remark this answers is not in the document]
Comment (Electronic-Waste) [reply hack/update-openapigen.sh]: SGTM. I've changed to use openapi binary.

File: hack/violation_exception_v1beta1.list [hunk 4]
```diff
@@ -0,0 +1,39 @@
+API rule violation: list_type_missing,github.com/kubeflow/katib/pkg/apis/controller/common/v1beta1,AlgorithmSpec,AlgorithmSettings
```
[the remark this answers is not in the document]
Comment (Electronic-Waste) [reply hack/violation_exception_v1beta1.list]: Yeah, I'll create an issue describing this after this PR get merged.

File: pkg/client/controller/applyconfiguration/common/v1beta1/algorithmsetting.go
Comment (helenxie-bit) [inline pkg/client/controller/applyconfiguration/common/v1beta1/algorithmsetting.go:1]: Are the files in `pkg/client/controller/applyconfiguration/common/v1beta1` no longer needed, or have they been replaced/generated in another location?

Comment (Electronic-Waste) [reply pkg/client/controller/applyconfiguration/common/v1beta1/algorithmsetting.go:1]: I guess, they are no longer needed since there are no `+genclient` annotations in `pkg/apis/controller/common/v1beta1`.

Comment (helenxie-bit) [reply pkg/client/controller/applyconfiguration/common/v1beta1/algorithmsetting.go:1]: Got it! That makes sense.

**Comment by helenxie-bit** [comment]
Thank you for the contribution! LGTM 🎉

File: .github/workflows/e2e-test-darts-cifar10.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (tenzen-y) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: I think that step by step upgrading would be better so that we can easily pick the commit when we find the issue on the specific K8s version.

So, I would recommend upgrading 1 K8s version in 1 PR.

File: go.mod [hunk 5]
```diff
@@ -14,29 +14,28 @@ require (
...
-	k8s.io/code-generator v0.29.3
-	k8s.io/klog v1.0.0
-	k8s.io/kube-openapi v0.0.0-20231010175941-2dd684a91f00
-	sigs.k8s.io/controller-runtime v0.17.3
+	google.golang.org/protobuf v1.34.1
+	k8s.io/api v0.30.7
+	k8s.io/apimachinery v0.30.7
+	k8s.io/client-go v0.30.7
+	k8s.io/code-generator v0.30.7
+	k8s.io/klog/v2 v2.130.1
```
Comment (tenzen-y) [inline go.mod:30]: Reverting klog version, then create dedicated issue for the klog version updating.

File: .github/workflows/e2e-test-darts-cifar10.yaml (hunk 1 above)
Comment (terrytangyuan) [inline .github/workflows/e2e-test-darts-cifar10.yaml:36]: We can also add one that tests whatever is the cutting edge version so we can get notified of any incompatibility earlier

Comment (Electronic-Waste) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: SGTM. WDYT👀 @kubeflow/wg-automl-leads

Comment (tenzen-y) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: Sorry for the late response. We upgrade the Kube version step by step.
So, I think that it would be great to use the 1.32 in separate issue.

I believe that @shaowei-su can open a PR to update Kube to 1.31 after this PR.
Then you can open PR to support 4th supporting Kube version, which meant that we support 1.29, 1.30, 1.31, 1.32 versions the same as training-operator.

Comment (andreyvelich) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: Yes, we discussed it today during the call.
Let's initially support 1.28 - 1.31 in Katib 0.18 as we discussed here: https://github.com/kubeflow/katib/issues/2457.
In the future, we can migrate to k8s 1.32.

Comment (Electronic-Waste) [reply .github/workflows/e2e-test-darts-cifar10.yaml:36]: SGTM. Can we work on this PR and merge it as soon as possible? The deadline for Kubeflow 1.10 feature freeze is Jan 20th. We may need to reserve some time for 1.31 version:)

File: Makefile [hunk 6]
```diff
@@ -82,7 +82,7 @@ sync-go-mod:
 CONTROLLER_GEN = $(shell pwd)/bin/controller-gen
 .PHONY: controller-gen
 controller-gen:
-	@GOBIN=$(shell pwd)/bin GO111MODULE=on go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.14.0
+	@GOBIN=$(shell pwd)/bin GO111MODULE=on go install sigs.k8s.io/controller-tools/cmd/controller-gen@v0.16.5
 
 # Run this if you update any existing controller APIs.
 # 1. Generate deepcopy, clientset, listers, informers for the APIs (hack/update-codegen.sh)
```
Comment (andreyvelich) [inline Makefile:88]: Please can we add the `go-mod-download` instruction to `make generate` similar to Training: https://github.com/kubeflow/training-operator/blob/master/Makefile#L50.

**Review by andreyvelich — COMMENTED** [review]
Thank you for this @Electronic-Waste
Overall lgtm.
/assign @tenzen-y @helenxie-bit @kubeflow/wg-training-leads

File: Makefile (hunk 6 above)
Comment (Electronic-Waste) [reply Makefile:88]: Got it! I will add it later.

**Review by andreyvelich — COMMENTED** [review]
Thank you @Electronic-Waste!
/lgtm
/approve
